File: Sources/Elastos/Frameworks/Droid/DevSamples/bak/Gallery/src/data/DataSourceHelper.cpp
#include "DataSourceHelper.h"
#include <elastos/utility/logging/Logger.h>
#include <elastos/core/StringUtils.h>

using Elastos::Core::StringUtils;
using Elastos::IO::CFile;
using Elastos::IO::IFile;
using Elastos::Utility::Logging::Logger;

namespace Gallery {

const String DataSourceHelper::ROOT_PATH = String("/mnt");
const String DataSourceHelper::PATH_SPLITE = String("/");
const String DataSourceHelper::SOURCE_PATH = String("source_path");
const String DataSourceHelper::SOURCE_DESC = String("source_desc");
const String DataSourceHelper::SOURCE_INDEX = String("source_index");
const String DataSourceHelper::TAG("DataSourceHelper");

Int32 DataSourceHelper::sVersion = 0;
String DataSourceHelper::sRootPath;
String DataSourceHelper::sCurrentPath;
Boolean DataSourceHelper::sInited = FALSE;

AutoPtr<List<String> > DataSourceHelper::sPath = new List<String>();
AutoPtr<HashMap<String, AutoPtr<List<String> > > > DataSourceHelper::sImageSourcePathMap
    = new HashMap<String, AutoPtr<List<String> > >();

DataSourceHelper::~DataSourceHelper()
{
    ClearAll();
}

void DataSourceHelper::SetRootPath(
    /* [in] */ const String& rootPath)
{
    Logger::D(TAG, "SetRootPath()---rootPath:%s", rootPath.string());
    ClearAll();
    sRootPath = rootPath;
}

AutoPtr<HashMap<String, AutoPtr<List<String> > > > DataSourceHelper::GetImageFileMap()
{
    Init();
    Logger::D(TAG, "GetImageFileMap()---count:%d", sImageSourcePathMap->GetSize());
    return sImageSourcePathMap;
}

AutoPtr<List<String> > DataSourceHelper::GetImagePathList()
{
    Init();
    Logger::D(TAG, "GetImagePathList()---count:%d", sPath->GetSize());
    if (sPath->GetSize() > 1) {
        SortListByTime();
    }
    return sPath;
}

AutoPtr<List<String> > DataSourceHelper::GetItemList(
    /* [in] */ const String& key)
{
    Init();
    AutoPtr<List<String> > list;
    HashMap<String, AutoPtr<List<String> > >::Iterator it = sImageSourcePathMap->Find(key);
    if (it != sImageSourcePathMap->End()) {
        list = it->mSecond;
    }

    Logger::D(TAG, "GetItemList()--key:%s, list:%p, mapSize:%d", key.string(), list.Get(), sImageSourcePathMap->GetSize());
    return list;
}

Boolean DataSourceHelper::TimeComparator(
    /* [in] */ const String& lhs,
    /* [in] */ const String& rhs)
{
    AutoPtr<IFile> file1;
    AutoPtr<IFile> file2;
    String strL(sCurrentPath);
    String strR(sCurrentPath);
    strL += lhs;
    strR += rhs;
    CFile::New(strL, (IFile**)&file1);
    CFile::New(strR, (IFile**)&file2);
    Int64 last1 = 0;
    Int64 last2 = 0;
    file1->LastModified(&last1);
    file2->LastModified(&last2);
    Int64 diff = last1 - last2;

    if (diff > 0) {
        return TRUE;
    }
    else if (diff == 0) {
        Int32 result = 0;
        if (sCurrentPath.IsNullOrEmpty()) {
            String lhsDir;
            String rhsDir;
            file1->GetParent(&lhsDir);
            file2->GetParent(&rhsDir);
            String lhsSub = lhsDir.Substring(lhsDir.LastIndexOf(PATH_SPLITE) + 1, lhsDir.GetLength());
            String rhsSub = rhsDir.Substring(rhsDir.LastIndexOf(PATH_SPLITE) + 1, rhsDir.GetLength());
            result = lhsSub.Compare(rhsSub);
        }
        else {
            result = lhs.Compare(rhs);
        }
        return result > 0 ? TRUE : FALSE;
    }
    else {
        return FALSE;
    }
}

Boolean DataSourceHelper::IsImageFile(
    /* [in] */ const String& fileName)
{
    // Logger::D(TAG, "IsImageFile--fileName:%s", fileName.string());
    Boolean result = FALSE;
    if (fileName.IsNullOrEmpty()) return result;
    String end = fileName.Substring(fileName.LastIndexOf(".") + 1, fileName.GetLength());
    end = end.ToLowerCase();
    if (end.Equals("jpg") || end.Equals("png") || end.Equals("jpeg") || end.Equals("bmp")
            || end.Equals("webp")) {
        result = TRUE;
    }
    return result;
}

void DataSourceHelper::Init()
{
    if (sInited)
        return;
    Logger::D(TAG, "Init()----");
    AutoPtr<IFile> dir;
    CFile::New(sRootPath, (IFile**)&dir);
    Boolean ret = FALSE;
    if (dir == NULL || (dir->IsDirectory(&ret), !ret)) {
        Logger::E(TAG, "sRootPath:%s is not a valid directory!", sRootPath.string());
        return;
    }

    GetImageFileMap(sRootPath);
    sInited = TRUE;
}

void DataSourceHelper::GetImageFileMap(
    /* [in] */ const String& rootPath)
{
    // Logger::D(TAG, "GetImageFileMap--rootPath:%s", rootPath.string());
    AutoPtr<IFile> dir;
    CFile::New(rootPath, (IFile**)&dir);
    AutoPtr< ArrayOf<IFile*> > files;
    dir->ListFiles((ArrayOf<IFile*>**)&files);
    if (files != NULL && files->GetLength() > 0) {
        AutoPtr<List<String> > itemList = new List<String>();
        Int32 length = files->GetLength();
        String absPath, name;
        IFile* file;
        Boolean ret;
        for (Int32 i = 0; i < length; i++) {
            file = (*files)[i];
            file->CanRead(&ret);
            if (ret) {
                file->IsDirectory(&ret);
                if (ret) {
                    file->GetAbsolutePath(&absPath);
                    GetImageFileMap(absPath);
                }
                else {
                    file->GetName(&name);
                    if (IsImageFile(name)) {
                        // Logger::D(TAG, "  > Image %d: %s", itemList->GetSize(), name.string());
                        itemList->PushBack(name);
                    }
                }
            }
        }

        if (itemList->IsEmpty() == FALSE) {
            Logger::D(TAG, "GetImageFileMap()--rootPath:%s, has %d images.", rootPath.string(), itemList->GetSize());
            if (itemList->GetSize() > 1) {
                sCurrentPath = rootPath;
                sCurrentPath += PATH_SPLITE;
                itemList->Sort(TimeComparator);
            }
            (*sImageSourcePathMap)[rootPath] = itemList;
            sPath->PushBack(rootPath);
        }
    }
}

void DataSourceHelper::ClearAll()
{
    Logger::D(TAG, "ClearAll()---");
    sPath->Clear();
    sImageSourcePathMap->Clear();
    sInited = FALSE;
}

void DataSourceHelper::SortListByTime()
{
    Logger::D(TAG, "SortListByTime()-----");
    AutoPtr<HashMap<String, AutoPtr<List<String> > > > pathMap = sImageSourcePathMap;
    AutoPtr<List<String> > pathList = sPath;
    List<String> albumList;
    AutoPtr<List<String> > fileList;
    List<String>::Iterator it = pathList->Begin();
    for (; it != pathList->End(); ++it) {
        String path(*it);
        HashMap<String, AutoPtr<List<String> > >::Iterator iter = pathMap->Find(path);
        fileList = iter->mSecond;
        path += PATH_SPLITE;
        path += fileList->GetFront();
        // Logger::D(TAG, "SortListByTime()-----path:%s", path.string());
        albumList.PushBack(path);
    }

    sCurrentPath = "";
    albumList.Sort(TimeComparator);
    Logger::D(TAG, "SortListByTime()----albumList-size:%d", albumList.GetSize());
    sPath->Clear();
    List<String>::Iterator iter = albumList.Begin();
    for (; iter != albumList.End(); ++iter) {
        String path(*iter);
        sPath->PushBack(path.Substring(0, path.LastIndexOf(DataSourceHelper::PATH_SPLITE)));
    }
    Logger::D(TAG, "SortListByTime()----sPath-size:%d", sPath->GetSize());
}

} // namespace Gallery

File: Sources/Elastos/Frameworks/Droid/DevSamples/jpk/Common/js/Gallery/data/DataSourceHelper.js
module.exports = function(aoElastos){

//common definition
var CString = aoElastos.Core.CString;
var Droid_New = aoElastos.Droid.New;
var Core_New = aoElastos.Core.New;

var DataSourceHelper =  {

//public:
SetRootPath : function (asRootPath) {
    this.ClearAll();
    this.sRootPath = asRootPath;
},
GetImageFileMap : function () {
    this.Init();
    return this.sImageSourcePathMap;
},
GetImagePathList : function () {
    this.Init();
    if (this.sPath.length > 1) {
        this.SortListByTime();
    }
    return this.sPath;
},
GetItemList : function (asKey) {
    this.Init();
    var ret = this.sImageSourcePathMap[asKey];
    return ret;
},
TimeComparator : function (lhs, rhs) {
    var strL = this.sCurrentPath + lhs;
    var strR = this.sCurrentPath + rhs;
    var file1 = Core_New('Elastos.IO.CFile', strL);
    var file2 = Core_New('Elastos.IO.CFile', strR);
    var last1 = file1.GetLastModified();
    var last2 = file2.GetLastModified();

    var diff = last1 - last2;

    if (diff > 0) return 1;
    if (diff < 0) return 0;

    var result = 0;
    if (this.sCurrentPath.length > 0) {
        //result = lhs.Compare(rhs);
        result = (lhs > rhs);
    }
    else {
        var lhsDir = file1.GetParent();
        var rhsDir = file2.GetParent();
        var lhsSub = lhsDir.split(this.PATH_SPLITE).pop();
        var rhsSub = rhsDir.split(this.PATH_SPLITE).pop();
        //result = lhs.Compare(rhs);
        result = (lhs > rhs);
    }
    var ret = result > 0 ? 1 : 0;

    return ret;
},
IsImageFile : function (asFileName) {
    var result = false;

    var a = asFileName.split('.');
    if (a.length < 2) {
        result = false;
    }
    else {
        var end = '.' + a.pop();
        result = '.jpg.png.jpeg.bmp.webp'.indexOf(end) >= 0;
    }

    return result;
},

//private:
Init : function () {
    if (this.sInited) return;

    var dir = Core_New('Elastos.IO.CFile', this.sRootPath);

    var ret = dir.IsDirectory();
    if (!ret) {
        elog("sRootPath:" + this.sRootPath + " is not a valid directory!");
    }

    this._GetImageFileMap(this.sRootPath);

    this.sInited = true;
},
_GetImageFileMap : function (rootPath) {
    var dir = Core_New('Elastos.IO.CFile', rootPath);
    var files = dir.ListFiles();
    if (!files || !files.length) return;

    var itemList = [];
    for (var i=0, im=files.length; i<im; i++) {
        var file = files[i];
        if (!file.CanRead()) continue;

        if (file.IsDirectory()) {
            var absPath = file.GetAbsolutePath();
            this._GetImageFileMap(absPath);
        }
        else {
            var name = file.GetName();
            if (name) itemList.push(name);
        }
    }

    if (itemList.length > 0) {
        this.sCurrentPath = rootPath + this.PATH_SPLITE;
        itemList.sort(this.TimeComparator.bind(this));

        this.sImageSourcePathMap[rootPath] = itemList;
        this.sPath.push(rootPath);
    }
},
ClearAll : function () {  //~DataSourceHelper()
    this.sPath = [];
    this.sImageSourcePathMap = {};
    this.sInited = false;
},
SortListByTime : function () {
    var pathMap = this.sImageSourcePathMap;
    var pathList = this.sPath;

    var albumList = [];
    for (var i=0, im=pathList.length; i<im; i++) {
        var path = pathList[i];
        var fileList = pathMap[path];
        path += this.PATH_SPLITE;
        path += fileList[0];
        albumList.push(path);
    }

    this.sCurrentPath = '';
    albumList.sort(this.TimeComparator.bind(this));
    this.sPath = [];
    for (var i = 0, im = albumList.length; i < im; i++) {
        var path = albumList[i];
        var aa = path.split(this.PATH_SPLITE);
        aa.pop();
        var end = aa.join(this.PATH_SPLITE);
        this.sPath.push(end);
    }
},

//public:
ROOT_PATH : "/data/temp/node/disk/testdisk",
PATH_SPLITE : "/",
SOURCE_PATH : "source_path",
SOURCE_DESC : "source_desc",
SOURCE_INDEX : "source_index",

//private:
TAG : "DataSourceHelper",
sVersion : 0,
sRootPath : '',
sCurrentPath : '',
sInited : false,

sPath : [], //['']
sImageSourcePathMap : {},   //{a:['b']}

};  //DataSourceHelper

DataSourceHelper.SetRootPath(DataSourceHelper.ROOT_PATH);

return DataSourceHelper;

};  //module.exports
